File: AIAndroid/app/src/main/kotlin/ai/ecoinference/app/AppState.kt
package ai.ecoinference.app

import android.content.Context
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import ai.ecoinference.app.inference.InferenceService
import ai.ecoinference.app.models.ModelCatalog
import ai.ecoinference.app.models.ModelInfo
import ai.ecoinference.app.services.DownloadService
import ai.ecoinference.app.services.SettingsService
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * Central application state — mirrors iOS AppState.swift.
 *
 * All UI state is exposed as [StateFlow]s. Methods mirror the iOS counterparts
 * so the Android and iOS code remain easy to keep in sync.
 */
class AppState(private val appContext: Context) : ViewModel() {

    companion object {
        fun factory(context: Context) = object : ViewModelProvider.Factory {
            @Suppress("UNCHECKED_CAST")
            override fun <T : ViewModel> create(modelClass: Class<T>): T =
                AppState(context.applicationContext) as T
        }
    }

    private val inference = InferenceService.getInstance(appContext)
    private val download  = DownloadService.getInstance(appContext)
    val settings          = SettingsService.getInstance(appContext)

    // ── Model state ───────────────────────────────────────────────────────────

    private val _modelLoaded        = MutableStateFlow(false)
    val modelLoaded: StateFlow<Boolean> = _modelLoaded.asStateFlow()

    private val _loadedModelId      = MutableStateFlow<String?>(null)
    val loadedModelId: StateFlow<String?> = _loadedModelId.asStateFlow()

    private val _isLoading          = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _loadError          = MutableStateFlow<String?>(null)
    val loadError: StateFlow<String?> = _loadError.asStateFlow()

    private val _loadErrorModelId   = MutableStateFlow<String?>(null)
    val loadErrorModelId: StateFlow<String?> = _loadErrorModelId.asStateFlow()

    /** True when the loaded model supports image attachment in the UI. */
    private val _imageInputEnabled  = MutableStateFlow(false)
    val imageInputEnabled: StateFlow<Boolean> = _imageInputEnabled.asStateFlow()

    // ── Download state ────────────────────────────────────────────────────────

    private val _downloadActive     = MutableStateFlow(false)
    val downloadActive: StateFlow<Boolean> = _downloadActive.asStateFlow()

    private val _downloadProgress   = MutableStateFlow(0f)
    val downloadProgress: StateFlow<Float> = _downloadProgress.asStateFlow()

    private val _downloadingModelId = MutableStateFlow<String?>(null)
    val downloadingModelId: StateFlow<String?> = _downloadingModelId.asStateFlow()

    private val _downloadError      = MutableStateFlow<String?>(null)
    val downloadError: StateFlow<String?> = _downloadError.asStateFlow()

    private val _downloadErrorModelId = MutableStateFlow<String?>(null)
    val downloadErrorModelId: StateFlow<String?> = _downloadErrorModelId.asStateFlow()

    // ── Catalog ───────────────────────────────────────────────────────────────

    private val _models             = MutableStateFlow<List<ModelInfo>>(emptyList())
    val models: StateFlow<List<ModelInfo>> = _models.asStateFlow()

    // ── Deep link ─────────────────────────────────────────────────────────────

    val deepLink = MutableStateFlow<DeepLinkAction?>(null)

    // ── Init ──────────────────────────────────────────────────────────────────

    init {
        refreshCatalog()
    }

    // ── Catalog ───────────────────────────────────────────────────────────────

    fun refreshCatalog() {
        val loadedId = inference.loadedModelId
        _models.value = ModelCatalog.all.map { info ->
            info.copy(
                downloaded = download.isDownloaded(info),
                loaded     = info.id == loadedId,
            )
        }
        _imageInputEnabled.value =
            _models.value.firstOrNull { it.loaded }?.supportsImageInput ?: false
    }

    // ── Download ──────────────────────────────────────────────────────────────

    fun startDownload(modelId: String, hfToken: String? = null) {
        val info = ModelCatalog.findById(modelId) ?: return
        if (_downloadActive.value) return

        _downloadActive.value       = true
        _downloadProgress.value     = 0f
        _downloadingModelId.value   = modelId
        _downloadError.value        = null
        _downloadErrorModelId.value = null

        viewModelScope.launch {
            try {
                val token = hfToken?.takeIf { it.isNotBlank() }
                    ?: settings.hfToken().takeIf { it.isNotBlank() }
                download.download(model = info, hfToken = token) { progress ->
                    _downloadProgress.value = (progress * 100).toFloat()
                }
                _downloadActive.value     = false
                _downloadProgress.value   = 100f
                _downloadingModelId.value = null
                refreshCatalog()
            } catch (e: Exception) {
                _downloadActive.value       = false
                _downloadError.value        = e.message ?: "Download failed"
                _downloadErrorModelId.value = modelId
                _downloadingModelId.value   = null
                refreshCatalog()
            }
        }
    }

    fun cancelDownload() {
        download.cancel()
    }

    fun deleteModel(modelId: String) {
        val info = ModelCatalog.findById(modelId) ?: return
        download.deleteModel(info)
        refreshCatalog()
    }

    // ── Model load / unload ───────────────────────────────────────────────────

    fun loadModel(modelId: String, useGpu: Boolean = false) {
        val info = ModelCatalog.findById(modelId) ?: return
        if (!download.isDownloaded(info)) return

        _isLoading.value        = true
        _loadError.value        = null
        _loadErrorModelId.value = null

        viewModelScope.launch {
            try {
                inference.load(
                    modelId      = modelId,
                    modelPath    = download.filePath(info).absolutePath,
                    useGpu       = useGpu,
                    maxNumTokens = info.maxContextTokens,
                    multimodal   = info.supportsVision,
                )
                _modelLoaded.value      = true
                _loadedModelId.value    = modelId
                _isLoading.value        = false
                _loadError.value        = null
                _loadErrorModelId.value = null
                refreshCatalog()
            } catch (e: Exception) {
                _loadError.value        = e.message ?: "Load failed"
                _loadErrorModelId.value = modelId
                _isLoading.value        = false
                refreshCatalog()
            }
        }
    }

    fun unloadModel() {
        inference.unload()
        _modelLoaded.value      = false
        _loadedModelId.value    = null
        _loadError.value        = null
        _loadErrorModelId.value = null
        _imageInputEnabled.value = false
        refreshCatalog()
    }
}

File: AIiOS/AIiOS/AppState.swift
import Foundation
import SwiftUI
import os.log
import FirebaseRemoteConfig

private let log = Logger(subsystem: "ai.ecoinference.app", category: "AppState")

// MARK: - Deep link actions

enum DeepLinkAction: Equatable {
    case openChat(prefill: String? = nil, autoSend: Bool = false, systemPrompt: String? = nil)
    case openModels
    case openSettings
    case loadModel(id: String)
    case backgroundInfer(prompt: String, callbackURL: URL?)
}

// MARK: - AppState

@MainActor
final class AppState: ObservableObject {

    static let shared = AppState()

    // ── Model ─────────────────────────────────────────────────────────────────

    @Published private(set) var modelLoaded        = false
    @Published private(set) var loadedModelId: String?
    @Published private(set) var isLoading          = false
    @Published private(set) var loadingModelId: String?
    @Published private(set) var loadError: String?
    @Published private(set) var loadErrorModelId: String?
    /// True when the loaded model's bundle supports image attachment in the UI.
    /// Distinct from the model being in multimodal mode (supportsVision) — a model
    /// may use the Conversation API without supporting actual image payloads.
    @Published private(set) var imageInputEnabled  = false

    // ── Download ──────────────────────────────────────────────────────────────

    @Published private(set) var downloadActive       = false
    @Published private(set) var downloadProgress: Double = 0   // 0–100
    @Published private(set) var downloadingModelId: String?
    @Published private(set) var downloadError: String?
    @Published private(set) var downloadErrorModelId: String?

    // ── Catalog ───────────────────────────────────────────────────────────────

    @Published private(set) var models: [ModelInfo] = []

    // ── Navigation / deep link ────────────────────────────────────────────────

    /// Set by the URL scheme handler; views react with .onChange(of: deepLink).
    @Published var deepLink: DeepLinkAction? = nil

    // ── Private ───────────────────────────────────────────────────────────────

    private let settings  = SettingsService.shared
    private let inference = InferenceService.shared
    private let download  = DownloadService.shared

    /// Set of model IDs enabled by Remote Config; nil = fetch pending (show all).
    private var enabledModelIds: Set<String>? = nil

    private init() {
        refreshCatalog()
        Task { await fetchRemoteConfig() }
    }

    private func fetchRemoteConfig() async {
        let rc = RemoteConfig.remoteConfig()
        rc.configSettings = RemoteConfigSettings()
        rc.configSettings.minimumFetchInterval = 3600
        do {
            try await rc.fetchAndActivate()
            let raw = rc.configValue(forKey: "available_models").stringValue
            guard let data = raw.data(using: .utf8),
                  let arr  = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]]
            else { return }
            let ids = Set(arr.compactMap { $0["id"] as? String }.filter { !$0.isEmpty })
            if !ids.isEmpty {
                enabledModelIds = ids
                refreshCatalog()
            }
        } catch {
            log.warning("Remote Config fetch failed: \(error.localizedDescription)")
        }
    }

    // MARK: - Catalog

    func refreshCatalog() {
        let loadedId  = inference.loadedModelId
        let allowlist = enabledModelIds
        models = ModelCatalog.all
            .filter { allowlist == nil || allowlist!.contains($0.id) }
            .map { info in
                var m = info
                m.downloaded = download.isDownloaded(m)
                m.loaded     = (m.id == loadedId)
                return m
            }
        imageInputEnabled = models.first(where: { $0.loaded })?.supportsImageInput ?? false
    }

    // MARK: - Download

    /// [authToken] is currently unused — model files are served unauthenticated from
    /// our own server. Kept so a future auth-locked server can pass a token through
    /// without changing this call site.
    func startDownload(modelId: String) {
        guard let info = ModelCatalog.find(id: modelId) else { return }
        guard !downloadActive else { return }

        downloadActive        = true
        downloadProgress      = 0
        downloadingModelId    = modelId
        downloadError         = nil
        downloadErrorModelId  = nil

        let downloadSvc = download

        Task {
            do {
                try await downloadSvc.download(model: info) { [weak self] progress in
                    Task { @MainActor [weak self] in
                        self?.downloadProgress = progress * 100
                    }
                }
                downloadActive     = false
                downloadProgress   = 100
                downloadingModelId = nil
                refreshCatalog()
            } catch {
                downloadActive       = false
                downloadError        = error.localizedDescription
                downloadErrorModelId = modelId
                downloadingModelId   = nil
                refreshCatalog()
            }
        }
    }

    func cancelDownload() {
        download.cancel()
    }

    func deleteModel(modelId: String) {
        guard let info = ModelCatalog.find(id: modelId) else { return }
        try? FileManager.default.removeItem(at: download.filePath(for: info))
        refreshCatalog()
    }

    // MARK: - Model load / unload

    func loadModel(modelId: String, useGpu: Bool = false) {
        guard let info = ModelCatalog.find(id: modelId),
              download.isDownloaded(info),
              !isLoading else { return }

        isLoading        = true
        loadingModelId    = modelId
        loadError        = nil
        loadErrorModelId = nil

        let inferenceSvc = inference
        let modelPath    = download.filePath(for: info).path

        Task.detached(priority: .userInitiated) { [weak self] in
            do {
                // flutter_gemma recommends 4096 tokens for multimodal models;
                // the vision KV-cache pushes memory usage high enough that 8192
                // causes litert_lm_engine_create to fail on device.
                try inferenceSvc.load(
                    modelId:             modelId,
                    modelPath:           modelPath,
                    useGpu:              useGpu,
                    maxNumTokens:        info.maxContextTokens,
                    multimodal:          info.supportsVision,
                    speculativeDecoding: info.supportsSpeculativeDecoding
                )
                await MainActor.run { [weak self] in
                    self?.modelLoaded       = true
                    self?.loadedModelId     = modelId
                    self?.isLoading         = false
                    self?.loadingModelId    = nil
                    self?.loadError         = nil
                    self?.loadErrorModelId  = nil
                    self?.refreshCatalog()
                    self?.deepLink          = .openChat()
                }
            } catch {
                await MainActor.run { [weak self] in
                    self?.loadError         = error.localizedDescription
                    self?.loadErrorModelId  = modelId
                    self?.isLoading         = false
                    self?.loadingModelId    = nil
                    self?.refreshCatalog()
                }
            }
        }
    }

    func unloadModel() {
        inference.unload()
        modelLoaded       = false
        loadedModelId     = nil
        loadError         = nil
        loadErrorModelId  = nil
        imageInputEnabled = false
        refreshCatalog()
    }
}
